heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs

```diff
@@ -3507,6 +3509,85 @@ remLLVMBLockPermRange rng bp =
+--
+-- where each @shi@ is equivalent up to associativity of the @;@ operator to a
+-- shape of the form
+--
+-- > fieldsh(eq(llvmword(bvi)));shi'
+--
+-- That is, each disjunct of the shape starts with an equality permission that
+-- determines which disjunct should be used. These shapes are represented as a
+-- list of the disjuncts, which are tagged with the bitvector values @bvi@ used
+-- in the equality permission.
+data TaggedUnionShape w
+  = forall sz. (1 <= sz, KnownNat sz) =>
+    TaggedUnionShape (NonEmpty (BV sz, PermExpr (LLVMShapeType w)))
+
+-- | Extract the disjunctive shapes from a 'TaggedUnionShape'
+taggedUnionDisjs :: TaggedUnionShape w -> [PermExpr (LLVMShapeType w)]
+taggedUnionDisjs (TaggedUnionShape disjs) =
+  snd $ unzip $ NonEmpty.toList disjs
+
+-- | Convert a 'TaggedUnionShape' to the shape it represents
+taggedUnionToShape :: TaggedUnionShape w -> PermExpr (LLVMShapeType w)
+taggedUnionToShape (TaggedUnionShape disjs) =
+  foldr1 PExpr_OrShape $ NonEmpty.map snd disjs
+
+-- | A bitvector value of some unknown size
+data SomeBV = forall sz. (1 <= sz, KnownNat sz) => SomeBV (BV sz)
+
+-- | Test if a shape is of the form @fieldsh(eq(llvmword(bv)))@ for some @bv@.
+-- If so, return @bv@.
+shapeToTag :: PermExpr (LLVMShapeType w) -> Maybe SomeBV
+shapeToTag (PExpr_FieldShape
+            (LLVMFieldShape
+             (ValPerm_Eq (PExpr_LLVMWord (PExpr_BV [] bv))))) =
+  Just (SomeBV bv)
+shapeToTag _ = Nothing
+
+-- | Test if a shape begins with an equality permission to a bitvector value and
+-- return that bitvector value
+getShapeBVTag :: PermExpr (LLVMShapeType w) -> Maybe SomeBV
+getShapeBVTag sh | Just some_bv <- shapeToTag sh = Just some_bv
+getShapeBVTag (PExpr_SeqShape sh1 _) = getShapeBVTag sh1
+getShapeBVTag _ = Nothing
+
+-- | Test if a shape is a tagged union shape and, if so, convert it to the
+-- 'TaggedUnionShape' representation
+asTaggedUnionShape :: PermExpr (LLVMShapeType w) -> Maybe (TaggedUnionShape w)
+asTaggedUnionShape (PExpr_OrShape sh1 sh2)
+  | Just (SomeBV bv1) <- getShapeBVTag sh1
+  , Just (TaggedUnionShape disjs2@((bv2,_) :| _)) <- asTaggedUnionShape sh2
+  , Just Refl <- testEquality (natRepr bv1) (natRepr bv2) =
+    Just (TaggedUnionShape (NonEmpty.cons (bv1,sh1) disjs2))
+asTaggedUnionShape sh
+  | Just (SomeBV bv) <- getShapeBVTag sh =
+    Just (TaggedUnionShape ((bv,sh) :| []))
+asTaggedUnionShape _ = Nothing
+
+-- | Find a disjunct in a 'TaggedUnionShape' that could be proven at the given
+-- offset from the given atomic permissions, by looking for a field or block
+-- permission containing an equality permission to one of the tags. If some
+-- disjunct can be proved, return its index in the list of disjuncts.
+findTaggedUnionIndexForPerms :: PermExpr (BVType w) ->
+                                [AtomicPerm (LLVMPointerType w)] ->
+                                TaggedUnionShape w -> Maybe Int
+findTaggedUnionIndexForPerms off (p : _) (TaggedUnionShape disjs@((bv1,_) :| _))
```

Comment:
This function currently appear to interleave the process of searching a list with matching a shape, maybe it would be more clear to break those up?

```haskell
findTaggedUnionIndexForPerms off ps sh =
  do TaggedUnionShape disjs@((bv1,_) :| _) <- Just sh
     asum (map findOne ps)
  where
    findOne p =
      do bp <- llvmAtomicPermToBlock p
         guard (bvEq off (llvmBlockOffset bp))
         SomeBV tag_bv <- getShapeBVTag (llvmBlockShape bp)
         Refl <- testEquality (natRepr tag_bv) (natRepr bv1)
         findIndex (\(i,_) -> i == tag_bv) (NonEmpty.toList disjs)
```

Reply:
This makes sense, thanks for the suggestion! My only change would be that I would want to make `findOne` its own top-level function, `findTaggedUnionIndexForPerm`, because of the documentation, correctness, and abstraction concerns I wrote about above.

Reply:
Making it top-level sounds great to me.